cpp/include/raft/spatial/knn/detail/warp_select_faiss.cuh

```diff
@@ -0,0 +1,739 @@
+#pragma unroll
+      for (int i = 0; i < kLowSize; ++i) {
+        k[i] = newK[i];
+        v[i].key = newV[i].key;
+        v[i].value = newV[i].value;
+      }
+    }
+
+    {
+      K newK[kHighSize];
+      KeyValuePair<K, V> newV[kHighSize];
+
+#pragma unroll
+      for (int i = 0; i < kHighSize; ++i) {
+        newK[i] = k[i + kLowSize];
+        newV[i].key = v[i + kLowSize].key;
+        newV[i].value = v[i + kLowSize].value;
+      }
+
+      constexpr bool kHighIsPowerOf2 =
+        utils::isPowerOf2(N - kNextHighestPowerOf2 / 2);
+      // FIXME: compiler doesn't like this expression? compiler bug?
+      //      constexpr bool kHighIsPowerOf2 = utils::isPowerOf2(kHighSize);
+      BitonicMergeStepKVP<K, V, kHighSize, Dir, Comp,
+                          false,  // high
+                          kHighIsPowerOf2>::merge(newK, newV);
+
+#pragma unroll
+      for (int i = 0; i < kHighSize; ++i) {
+        k[i + kLowSize] = newK[i];
+        v[i + kLowSize].key = newV[i].key;
+        v[i + kLowSize].value = newV[i].value;
+      }
+    }
+  }
+};
+
+/// Merges two sets of registers across the warp of any size;
+/// i.e., merges a sorted k/v list of size kWarpSize * N1 with a
+/// sorted k/v list of size kWarpSize * N2, where N1 and N2 are any
+/// value >= 1
+template <typename K, typename V, int N1, int N2, bool Dir, typename Comp,
+          bool FullMerge = true>
+inline __device__ void warpMergeAnyRegistersKVP(K k1[N1],
+                                                KeyValuePair<K, V> v1[N1],
+                                                K k2[N2],
+                                                KeyValuePair<K, V> v2[N2]) {
+  constexpr int kSmallestN = N1 < N2 ? N1 : N2;
+
+#pragma unroll
+  for (int i = 0; i < kSmallestN; ++i) {
+    K& ka = k1[N1 - 1 - i];
+    KeyValuePair<K, V>& va = v1[N1 - 1 - i];
+
+    K& kb = k2[i];
+    KeyValuePair<K, V>& vb = v2[i];
+
+    K otherKa;
+    KeyValuePair<K, V> otherVa;
+
+    if (FullMerge) {
+      // We need the other values
+      otherKa = shfl_xor(ka, kWarpSize - 1);
+      K otherVak = shfl_xor(va.key, kWarpSize - 1);
+      V otherVav = shfl_xor(va.value, kWarpSize - 1);
+      otherVa = KeyValuePair(otherVak, otherVav);
+    }
+
+    K otherKb = shfl_xor(kb, kWarpSize - 1);
+    K otherVbk = shfl_xor(vb.key, kWarpSize - 1);
+    V otherVbv = shfl_xor(vb.value, kWarpSize - 1);
+
+    // ka is always first in the list, so we needn't use our lane
+    // in this comparison
+    bool swapa = Dir ? Comp::gt(ka, otherKb) : Comp::lt(ka, otherKb);
+    assign(swapa, ka, otherKb);
+    assign(swapa, va.key, otherVbk);
+    assign(swapa, va.value, otherVbv);
+
+    // kb is always second in the list, so we needn't use our lane
+    // in this comparison
+    if (FullMerge) {
+      bool swapb = Dir ? Comp::lt(kb, otherKa) : Comp::gt(kb, otherKa);
+      assign(swapb, kb, otherKa);
+      assign(swapb, vb.key, otherVa.key);
+      assign(swapb, vb.value, otherVa.value);
+
+    } else {
```

Comment:
Any reason to have this empty else clause here as opposed to just putting a comment below?

Reply:
This code is actually taken directly from FAISS and modified. I tried to limit the modifications so that we can more easily track them and update if necessary. I'm also hoping the high dims PR might be able to remove this altogether.

Reply:
Great! That's more than sufficient reason to leave it as is.